lf-queue/lfq.c

```diff
@@ -0,0 +1,214 @@
+                return i;
+        }
+    }
+
+    return -1;
+}
+
+static void free_tid(struct lfq_ctx *ctx, int tid)
+{
+    ctx->tid_map[tid] = 0;
+}
+
+int lfq_init(struct lfq_ctx *ctx, int max_consume_thread)
+{
+    struct lfq_node *tmp = calloc(1, sizeof(struct lfq_node));
+    if (!tmp)
+        return -errno;
+
+    struct lfq_node *node = calloc(1, sizeof(struct lfq_node));
+    if (!node)
+        return -errno;
+
+    tmp->can_free = node->can_free = true;
+    memset(ctx, 0, sizeof(struct lfq_ctx));
+    ctx->MAX_HP_SIZE = max_consume_thread;
+    ctx->HP = calloc(max_consume_thread, sizeof(struct lfq_node));
+    ctx->tid_map = calloc(max_consume_thread, sizeof(struct lfq_node));
+    ctx->head = ctx->tail = tmp;
+    ctx->fph = ctx->fpt = node;
+
+    return 0;
+}
+
+long lfg_count_freelist(const struct lfq_ctx *ctx)
+{
+    long count = 0;
+    for (struct lfq_node *p = (struct lfq_node *) ctx->fph; p; p = p->free_next)
+        count++;
+    return count;
+}
+
+int lfq_release(struct lfq_ctx *ctx)
+{
+    if (ctx->tail && ctx->head) {               /* if we have data in queue */
+        while ((struct lfq_node *) ctx->head) { /* while still have node */
+            struct lfq_node *tmp = (struct lfq_node *) ctx->head->next;
+            safe_free(ctx, (struct lfq_node *) ctx->head);
+            ctx->head = tmp;
+        }
+        ctx->tail = 0;
+    }
+    if (ctx->fph && ctx->fpt) {
+        free_pool(ctx, true);
+        if (ctx->fph != ctx->fpt)
+            return -1;
+        free(ctx->fpt); /* free the empty node */
+        ctx->fph = ctx->fpt = 0;
+    }
+    if (ctx->fph || ctx->fpt)
+        return -1;
+
+    free(ctx->HP);
+    free(ctx->tid_map);
+    memset(ctx, 0, sizeof(struct lfq_ctx));
+
+    return 0;
+}
+
+int lfq_enqueue(struct lfq_ctx *ctx, void *data)
+{
+    struct lfq_node *insert_node = calloc(1, sizeof(struct lfq_node));
+    if (!insert_node)
+        return -errno;
+
+    insert_node->data = data;
+    struct lfq_node *old_tail = XCHG(&ctx->tail, insert_node);
+    /* We have claimed our spot in the insertion order by modifying tail.
+     * we are the only inserting thread with a pointer to the old tail.
+     *
+     * Now we can make it part of the list by overwriting the NULL pointer in
+     * the old tail. This is safe whether or not other threads have updated
+     * ->next in our insert_node.
+     */
+    assert(!(old_tail->next) && "old tail was not NULL");
+    atomic_store(&old_tail->next, insert_node);
+
+    return 0;
+}
+
+void *lfq_dequeue_tid(struct lfq_ctx *ctx, int tid)
+{
+    struct lfq_node *old_head, *new_head;
+
+    /* HP[tid] is necessary for deallocation. */
+    do {
+    retry:
+        /* continue jumps to the bottom of the loop, and would attempt a CAS
+         * with uninitialized new_head.
+         */
+        old_head = atomic_load(&ctx->head);
+
+        /* seq-cst store.
+         * FIXME: use xchg instead of mov + mfence on x86.
```

Comment:
Validate this comment for the sake of C11 atomics transition.

Reply:
It is not valid. Removed.